### DISCUSSION TITLE
how to disconnect (kick) client from server end?

### DISCUSSION THREAD

**Discussion by sluofoss** [miguelgrinberg/python-socketio#838] [Q&A] [answered]
I'm looking at the [doc](https://python-socketio.readthedocs.io/en/latest/server.html#connect-and-disconnect-event-handlers) and it seems like this disconnect function is triggered only when client initiates a disconnect. What is the python-socketio's implementation of [server.disconnectSockets](https://socket.io/docs/v4/server-api/#serverdisconnectsocketsclose)?

More on this, if server can't proactively kick sockets, how does it prevent multiple clients spamming connections?

**Comment by miguelgrinberg** [comment, marked as answer]
There is no method to disconnect all clients in a single call, but you can [disconnect a client](https://python-socketio.readthedocs.io/en/latest/api.html#socketio.Server.disconnect), given its `sid`.

**Reply by sluofoss** [reply]
Hi, im really sorry about this, is there a minimal working example of client server messaging with emit and disconnect? I tried adding on the async client 

```
await sio.emit("my_message",'first message') 
```

and there is no print on either side.


```
import asyncio
import socketio

sio = socketio.AsyncClient()

@sio.event
async def connect():
    print('connection established')

@sio.event
async def my_message(data):
    print('message received with ', data)
    await sio.emit('my response', {'response': 'my response'})

@sio.event
async def disconnect():
    print('disconnected from server')

async def main():
    await sio.connect('http://localhost:8080')
    await sio.emit('my_message',"haha")
    await sio.wait()

if __name__ == '__main__':
    asyncio.run(main())
``` 

Edit:

below is my attempt on disconnecting on the non async version

```
sio = socketio.Server()
app = socketio.WSGIApp(sio, static_files={
    '/': {'content_type': 'text/html', 'filename': 'index.html'}
})

clientDict = {}
@sio.event
def connect(sid, environ):
    print('connect ', sid)
    clientDict[sid] = 0
    sio.emit("MY EVENT","1st message",room =sid)
    sio.disconnect(sid)
@sio.event
def my_message(sid, data):
    print('message ', data)

@sio.event
def disconnect(sid):
    print('disconnect ', sid)

if __name__ == '__main__':
    eventlet.wsgi.server(eventlet.listen(('', 5000)), app)
```

**Reply by miguelgrinberg** [reply]
There is really not a lot of justification to use Socket.IO if all you want to do is emit and disconnect. An HTTP request is a much better option for that use case.

But in any case, you are trying to emit and disconnect in the connect event handler. This handler is used to authenticate the client, so at this point the connection isn't fully established. If you want to accept the client you have to exit the connect handler and do your emit in a different handler or in a background task that runs after the connection has been accepted.

The second mistake that you made is to call emit and then disconnect right after. The emits are asynchronous, when the emit returns the client may not have received the event yet. You can use the `callback` feature of the emit call, or else use the `call()` method instead of `emit()` to wait for the emit to be received and acknowledged. Only at that point it is safe to call disconnect.

**Reply by sluofoss** [reply]
Thank you so much! What i'm hoping to achieve is to have each client receive 100 emits from the server at arbitrary interval and after this limit is reached, the server forcefully closes the connection. So I guess what I need is use ```call``` to ensure the client received the message, and set up an async counter outside that is observed to trigger disconnect.
